Respect --set=target.platform when building rustbuild itself

MarcusCalhoun-Lopez opened the pull request:
`--set=target.platform.cc` and `--set=target.platform.cxx` are ignored if target is quoted.

`--set=target.platform.linker` is ignored if RUSTFLAGS is not set.

Undo parts of
https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488 and
https://github.com/rust-lang/rust/commit/1532fd8cd0db93f469e414f9da31ef083a44fcba

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -753,6 +753,9 @@ def build_bootstrap(self, color):
             target_features += ["-crt-static"]
         if target_features:
             env["RUSTFLAGS"] += " -C target-feature=" + (",".join(target_features))
+        target_linker = self.get_toml("linker", build_section)
+        if target_linker is not None:
+            env["RUSTFLAGS"] += " -C linker=" + target_linker
```
Why is this necessary? Don't we set it above on line 745?

MarcusCalhoun-Lopez replied:
Thank you for looking at this pull request.

In the MacPorts project, it is important to be able to [specify the compiler and linker](https://trac.macports.org/wiki/UsingTheRightCompiler).
To accomplish this, the [MacPorts recipe](https://github.com/macports/macports-ports/blob/85118394f6742209e4daf000961fcfcffe4ef810/lang/rust/Portfile#L141) uses
`--set=target.${platform}.cc` and `--set=target.${platform}.linker`.

However, `[target.'x86_64-apple-darwin']` is not recognized in config.toml.
Whereas `[target.x86_64-apple-darwin]` works fine.

Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.

Plese forgive me if I am missing something obvious.

jyn514 replied:
> However, `[target.'x86_64-apple-darwin']` is not recognized in config.toml.
> Whereas `[target.x86_64-apple-darwin]` works fine.

Can you post the error you see that makes you think this isn't recognized? This is valid toml syntax and the two should be exactly the same:
```sh
$  ./configure --set target.x86_64-apple-darwin.cc=/not/a/real/compiler
$ grep 'cc = .*/not' config.toml -B6
[target.'x86_64-apple-darwin']

# C compiler to be used to compile C code. Note that the
# default value is platform specific, and if not specified it may also depend on
# what platform is crossing to what platform.
# See `src/bootstrap/cc_detect.rs` for details.
cc = '/not/a/real/compiler'
$ toml get config.toml target.x86_64-apple-darwin.cc
"/not/a/real/compiler"
```

And if we merge your change it will break a previous bug, so I don't really want to land it as-is.

> Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.

What do you mean by "not recognized"? What error are you seeing?

jyn514 replied:
Ok, I think I see the issue - bootstrap.py is parsing config.toml incorrectly:
```
$ git diff
diff --git a/src/bootstrap/bootstrap_test.py b/src/bootstrap/bootstrap_test.py
index 06ca3ce21b3..8dec0de03df 100644
--- a/src/bootstrap/bootstrap_test.py
+++ b/src/bootstrap/bootstrap_test.py
@@ -73,6 +73,13 @@ class ProgramOutOfDate(unittest.TestCase):
             rustc_stamp.write("2017-06-15None")
         self.assertFalse(self.build.program_out_of_date(self.rustc_stamp_path, self.key))
 
+class ParseToml(unittest.TestCase):
+    def test_parse_quotes(self):
+        rb = bootstrap.RustBuild()
+        rb.config_toml = "[build]\ncc = 'xyz'\n[target.'x86_64-apple-darwin']\ncc = '/not/a/real/compiler'"
+        self.assertEqual(rb.get_toml("cc", "build"), "xyz")
+        cc = rb.get_toml("cc", "target.x86_64-apple-darwin")
+        self.assertEqual(cc, "/not/a/real/compiler")
 
 if __name__ == '__main__':
     SUITE = unittest.TestSuite()
@@ -80,6 +87,7 @@ if __name__ == '__main__':
     SUITE.addTest(doctest.DocTestSuite(bootstrap))
     SUITE.addTests([
         TEST_LOADER.loadTestsFromTestCase(VerifyTestCase),
+        TEST_LOADER.loadTestsFromTestCase(ParseToml),
         TEST_LOADER.loadTestsFromTestCase(ProgramOutOfDate)])
 
     RUNNER = unittest.TextTestRunner(stream=sys.stdout, verbosity=2)
$ x test bootstrap
======================================================================
FAIL: test_parse_quotes (__main__.ParseToml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyn/src/rust/rust/src/bootstrap/bootstrap_test.py", line 82, in test_parse_quotes
    self.assertEqual(cc, "/not/a/real/compiler")
AssertionError: None != '/not/a/real/compiler'
```
Let's fix that directly instead of removing the quotes.

MarcusCalhoun-Lopez replied:
> And if we merge your change it will break a previous bug, so I don't really want to land it as-is.

I added another attempt to fix the problem that should not break a previous bug.

> > Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.
> 
> What do you mean by "not recognized"? What error are you seeing?

Sorry if I was not clear.
I mean the linker set in `--set=target.platform.linker` is **not** used unless RUSTFLAGS is set as was done before the [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3).

> Please also add the test case I posted here.

I added the test case as requested.

jyn514 replied:
> Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting LD_x86_64_apple_darwin does not seem to be recognized whereas setting RUSTFLAGS does.

Does it work for `target.x86_64-apple-darwin`? Or is it consistently broken for both `target.x86_64-apple-darwin` and `target.'x86_64-apple-darwin'?

jyn514 reviewed (commented):
Can you explain more why this change is necessary? Won't this break https://github.com/rust-lang/rust/issues/97263 again?

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3776523743/jobs/6419839558) [(plain)](https://github.com/rust-lang/rust/commit/f8521b3e96f20473b5e3b9c7c77b32f35dd9fdbb/checks/10295657983/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built c89e441075fe
Successfully tagged rust-ci:latest
Built container sha256:c89e441075fe933fe7a968a3c9125db5e6f179838bc8b8b23f98242a3941f130
Uploading finished image to https://ci-caches.rust-lang.org/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45
upload failed: - to s3://rust-lang-ci-sccache2/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
Found 507 error codes
Found 0 error(s) in error codes
Done!
* 392 features
tidy error: /checkout/src/bootstrap/bootstrap_test.py:79: line longer than 100 chars
some tidy checks failed
```

</details>

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -734,7 +734,7 @@ def build_bootstrap(self, color):
 
         # Export Stage0 snapshot compiler related env variables
-        build_section = "target.{}".format(self.build)
+        build_section = "target.'{}'".format(self.build)
```
This isn't the correct fix, this will break ,`target.x86_64-unknown-linux-gnu` (without the quotes). Please add that as a test case and fix this in `get_toml`, not at the call site.

MarcusCalhoun-Lopez replied:
As requested, I attempted the fix in `get_toml`.
Thank you for your patience.

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -629,6 +629,12 @@ def get_toml(self, key, section=None):
         'true'
         """
 
+        # Target is quoted in bootstrap configure.
+        # See https://github.com/rust-lang/rust/commit/1532fd8cd0db93f469e414f9da31ef083a44fcba
+        # However, we do not want to break target.x86_64-unknown-linux-gnu (without the quotes).
+        if section is not None and section.startswith("target.") and section != "target.x86_64-unknown-linux-gnu":
+            section = "target.'{}'".format(section[7:])
```
This is not quite what I meant - you've hard-coded the specific case I called out earlier, but I want to fix it generally, not just for `target.x86_64-unknown-linux-gnu`. The proper fix would be some change to `section_match` to allow quotes around the section; maybe just `.replace("'", "")`? at this point I wonder if we should bring in a toml parser, but `configparse` doesn't work and anything else requires us to pull in a dependency :/

cc @Mark-Simulacrum, we're parsing .toml incorrectly in python because of the regex and I was wondering if you had suggestions for how to fix it.

jyn514 replied:
I guess we could vendor [tomllib](https://docs.python.org/3/library/tomllib.html) into bootstrap although I'm not super excited about that.

Mark-Simulacrum replied:
I don't think we should vendor anything into our code.

Can you say more about "target is quoted" from the PR description? In particular, if the problem is that we don't support a particular syntax but there are workarounds, that doesn't really worry me.

In general, I think `--set` should be considered best-effort; users who want full fidelity can always directly append custom TOML to config.toml without too much trouble IMO. If we want to further support that I would be OK with a `--append file/path` option that inlines some other file(s) as well or similar, but I'm not sure it's really necessary.

jyn514 replied:
@Mark-Simulacrum some history: *we ourselves* are the ones adding quotes to the target name in `configure`: https://github.com/rust-lang/rust/blob/4151a5b96cb4707bb0b09037bacc58ade1940fa2/src/bootstrap/configure.py#L408
The reason for that was to allow targets that had a `.` in their name: https://github.com/rust-lang/rust/issues/97263
So we're generating configuration in `configure.py` that `bootstrap.py` can't read, which seems pretty bad to me.

The reason we didn't notice until now is that *only* bootstrap.py couldn't read it; the toml parsing is correct in rustbuild. I think @MarcusCalhoun-Lopez ran into issues because he's configuring the linker for a target and expecting that to be used for *rustbuild itself*, which was the case until we started quoting targets.

> In general, I think --set should be considered best-effort

I am not sure that's relevant here, since even modifying config.toml manually won't get us the right behavior; you'll still be unable to use quotes around a target name and have bootstrap.py parse it correctly.

Mark-Simulacrum replied:
Yes, that does seem different (and talking about configure/--set doesn't seem entirely relevant, as you say). It sounds like we're always generating valid TOML, we just can't parse it later?

That said, it also seems like a valid workaround *is* generating TOML that bootstrap.py is capable of parsing, right? (Perhaps so long as your target's name doesn't contain `.`).

I think long-term https://docs.python.org/3/library/tomllib.html is the way to go; if we don't remove bootstrap.py's need for parsing complex TOML via the RIIR effort for rustbuild.

It will presumably be available on most distros in the next year or three...

Mark-Simulacrum replied:
I'd also be okay (if this works) not quoting things that don't contain `.`. maybe that fixes our parsing too?

jyn514 replied:
> I'd also be okay (if this works) not quoting things that don't contain `.`. maybe that fixes our parsing too?

:+1:  that seems like a reasonable compromise in the meantime.

MarcusCalhoun-Lopez replied:
I have updated the pull request that I believe addresses the concerns.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3777005330/jobs/6420599469) [(plain)](https://github.com/rust-lang/rust/commit/738c5f537db7ab85821b9e8c27541f3e48f1ee3f/checks/10296518182/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 8c0b3a0beb15
Successfully tagged rust-ci:latest
Built container sha256:8c0b3a0beb156b7a74d630d8ec71b9325062ed0475f267fd610c6df65d8a0b16
Uploading finished image to https://ci-caches.rust-lang.org/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45
upload failed: - to s3://rust-lang-ci-sccache2/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
Found 507 error codes
Found 0 error(s) in error codes
Done!
* 392 features
tidy error: /checkout/src/bootstrap/bootstrap_test.py:79: line longer than 100 chars
tidy error: /checkout/src/bootstrap/bootstrap.py:635: line longer than 100 chars
some tidy checks failed
```

</details>

MarcusCalhoun-Lopez committed:
Respect --set=target.platform during build

Avoid quoting targets that do not contain a period.
See https://github.com/rust-lang/rust/commit/1532fd8cd0db93f469e414f9da31ef083a44fcba

`--set=target.platform.linker` is ignored if RUSTFLAGS is not set.
Undo parts of https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3778940190/jobs/6423805988) [(plain)](https://github.com/rust-lang/rust/commit/86040d23ff66c17c53858e9975777d91fc213ba1/checks/10300192380/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 56ae7b3869fb
Successfully tagged rust-ci:latest
Built container sha256:56ae7b3869fb7a361b1a1aa66ae8f87dd303b43532e2723f1ae7b64afa0c6c56
Uploading finished image to https://ci-caches.rust-lang.org/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45
upload failed: - to s3://rust-lang-ci-sccache2/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
Found 507 error codes
Found 0 error(s) in error codes
Done!
* 392 features
tidy error: /checkout/src/bootstrap/bootstrap_test.py:79: line longer than 100 chars
some tidy checks failed
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3782319273/jobs/6429961857) [(plain)](https://github.com/rust-lang/rust/commit/1a5d6e4b2fcb377991a8aeffdcc6241839abf91d/checks/10307558336/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 5b92c2c50335
Successfully tagged rust-ci:latest
Built container sha256:5b92c2c5033525644843bf6b6764eb36ecbedbb603d9e8377561d8f36c757659
Uploading finished image to https://ci-caches.rust-lang.org/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45
upload failed: - to s3://rust-lang-ci-sccache2/docker/58bd69d3781a382eb5941028d9596b84692bd50bf9ab53a2d16b350022deca32f83037b6d64da3bfdf69e2b275e0db2ddd3212390c690549c0b3382192828b45 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
Found 507 error codes
Found 0 error(s) in error codes
Done!
* 392 features
tidy error: /checkout/src/bootstrap/bootstrap_test.py:79: line longer than 100 chars
some tidy checks failed
```

</details>

jyn514 commented on src/bootstrap/bootstrap_test.py:
```diff
@@ -73,13 +73,21 @@ def test_same_dates(self):
             rustc_stamp.write("2017-06-15None")
         self.assertFalse(self.build.program_out_of_date(self.rustc_stamp_path, self.key))
 
+class ParseToml(unittest.TestCase):
+    def test_parse_quotes(self):
+        rb = bootstrap.RustBuild()
+        rb.config_toml = "[build]\ncc = 'xyz'\n[target.'x86_64-apple-darwin']\ncc = '/not/a/real/compiler'"
+        self.assertEqual(rb.get_toml("cc", "build"), "xyz")
+        cc = rb.get_toml("cc", "target.x86_64-apple-darwin")
+        self.assertEqual(cc, "/not/a/real/compiler")
```
This test is going to fail now, since you haven't changed `get_toml`. It's ok to remove it since we're working around the problem in a different way.

jyn514 commented on src/bootstrap/configure.py:
```diff
@@ -405,7 +405,7 @@ def set(key, value):
 for target in configured_targets:
     targets[target] = sections['target'][:]
-    targets[target][0] = targets[target][0].replace("x86_64-unknown-linux-gnu", "'{}'".format(target))
+    targets[target][0] = targets[target][0].replace("x86_64-unknown-linux-gnu", "'{}'".format(target) if "." in target else target)
```
Please leave a comment here explaining why `.` is special-cased.
```suggestion
    # For `.` to be valid TOML, it needs to be quoted. But `bootstrap.py` doesn't use a proper TOML parser and fails to parse the target.
    # Avoid using quotes unless it's necessary.
    targets[target][0] = targets[target][0].replace("x86_64-unknown-linux-gnu", "'{}'".format(target) if "." in target else target)
```

MarcusCalhoun-Lopez replied:
Thank you for the wording.
The comments have been added to the pull request.

jyn514 reviewed (commented):
This popped up again, for a target that *does* have a dot in it and wouldn't be fixed by this PR ... it makes me little uncomfortable to keep piling up hacks instead of fixing the parsing in bootstrap.py. https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/specifying.20linker.20for.20cross-compiling

MarcusCalhoun-Lopez commented:
> This popped up again, for a target that _does_ have a dot in it and wouldn't be fixed by this PR ... it makes me little uncomfortable to keep piling up hacks instead of fixing the parsing in bootstrap.py. https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/specifying.20linker.20for.20cross-compiling

I am afraid I do not know how to implement a "proper" fix.

jyn514 commented:
Ok, this seems reasonable. We can worry about a permanent fix later, this is still better than the status quo.

@MarcusCalhoun-Lopez thank you for sticking with this :)

@bors r+

bors commented:
:pushpin: Commit 480297d216cbaad9527b7004222cd6c1faaa5e74 has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 480297d216cbaad9527b7004222cd6c1faaa5e74 -->
<!-- homu: {"type":"Approved","sha":"480297d216cbaad9527b7004222cd6c1faaa5e74","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 480297d216cbaad9527b7004222cd6c1faaa5e74 with merge 2c7536eaae59b44ff5259183ca98384129d69dd0...
<!-- homu: {"type":"BuildStarted","head_sha":"480297d216cbaad9527b7004222cd6c1faaa5e74","merge_sha":"2c7536eaae59b44ff5259183ca98384129d69dd0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3803424157/jobs/6471009572)
Approved by: jyn514
Pushing 2c7536eaae59b44ff5259183ca98384129d69dd0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3803424157/jobs/6471009572"},"merge_sha":"2c7536eaae59b44ff5259183ca98384129d69dd0"} -->

rust-timer commented:
Finished benchmarking commit (2c7536eaae59b44ff5259183ca98384129d69dd0): [comparison URL](https://perf.rust-lang.org/compare.html?start=ad8ae0504c54bc2bd8306abfcfe8546c1bb16a49&end=2c7536eaae59b44ff5259183ca98384129d69dd0&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ad8ae0504c54bc2bd8306abfcfe8546c1bb16a49&end=2c7536eaae59b44ff5259183ca98384129d69dd0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.0%, -1.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.7% | [-3.7%, -3.7%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-1.0%, -1.0%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ad8ae0504c54bc2bd8306abfcfe8546c1bb16a49&end=2c7536eaae59b44ff5259183ca98384129d69dd0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3% | [2.2%, 2.3%] | 2     |
| Regressions ❌ <br /> (secondary)  | 6.9% | [6.9%, 6.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.3% | [2.2%, 2.3%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ad8ae0504c54bc2bd8306abfcfe8546c1bb16a49&end=2c7536eaae59b44ff5259183ca98384129d69dd0&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
